feat: add experimental `@sveltejs/image` package (#9787)
sveltejs/kit: packages/image/README.md

# `@sveltejs/image`

> **WARNING**: This package is experimental. It uses pre-1.0 versioning and may introduce breaking changes with every minor version release.

Author: We should definitely do that but only once we feel more confident in the API. Right now I want this to be a "oh interesting I want to try this out" thing.

> This package aims to bring a plug&play image component to SvelteKit that is opinionated enough so you don't have to worry about the details, yet flexible enough for more advanced use cases or tweaks. It uses the `srcset` and `sizes` attributes of the `img` tag to provide resized images suitable for various device sizes, which for example results in smaller images downloaded for mobile.

Reviewer: ```suggestion
This package aims to bring a plug and play image component to SvelteKit that is opinionated enough so you don't have to worry about the details, yet flexible enough for more advanced use cases or tweaks. It uses the `srcset` and `sizes` attributes of the `img` tag to provide resized images suitable for various device sizes, which for example results in smaller images downloaded for mobile.
```

## Setup

Install:

```bash
npm install --save @sveltejs/image
```

Adjust `vite.config.js`:

```diff
+import { vitePluginSvelteImage } from '@sveltejs/image/vite';
import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vite';

export default defineConfig({
	plugins: [

> +		vitePluginSvelteImage({ providers: { default: '@sveltejs/image/providers/<choose one>' } }),

Reviewer: can we call this something shorter? the name looks really different than the `sveltekit()` plugin on the line below. how about just `images()`

		sveltekit()
	]
});
```

> > `<choose one>` refers to chosing one of the ready-to-use providers. We plan to add more providers over time. You can create your own by creating a JavaScript with a `export function getURL({ src, width, height }): string` function inside.

Reviewer: we could make this compatible with `unpic` with a small tweak, which would add over a dozen new providers. (also there's a typo in "chosing" -> "choosing")
```suggestion
> `<choose one>` refers to choosing one of the ready-to-use providers. We plan to add more providers over time. You can create your own by creating a JavaScript with a `export function transformUrl({ url, width, height }): string` function inside.
```

> > `<choose one>` refers to chosing one of the ready-to-use providers. We plan to add more providers over time. You can create your own by creating a JavaScript with a `export function getURL({ src, width, height }): string` function inside.

Reviewer: ```suggestion
> `<choose one>` refers to choosing one of the ready-to-use providers. We plan to add more providers over time. You can create your own by creating a JavaScript with a `export function getURL({ src, width, height }): string` function inside.
```

In case of Vercel, adjust `svelte.config.js`:

```diff
import adapter from '@sveltejs/adapter-vercel';

/** @type {import('@sveltejs/kit').Config} */
const config = {
	kit: {
-		adapter: adapter()
+		adapter: adapter({ images: true })
	}
};

export default config;
```

## Usage

### When using one of the providers, i.e. an image CDN:

```svelte
<script>
	import Image from '@sveltejs/image';
</script>

> <Image src="/path/to/your/image.jpg" width={1200} height={1800} alt="An alt text" />

Reviewer: This cannot be cached. If you `import` it then Vite will compute a hash giving it a unique URL that you can set far forwards expires headers one. I think we should still encourage images to be imported for dynamic CDNs.

You could set a project-level default for static vs dynamic and then opt-in/out with a query parameter that triggers the `include`/`exclude` of `vite-imagetools`

Author: ".. that you can set far forwards expires headers one" - I don't understand what you mean by that

Reviewer: "one" was a typo and was supposed to be "on". What I mean is setting `'cache-control': 'public, immutable, max-age=31536000'`

Reviewer: When we talked this morning you said `/path/to/your/image.jpg` is something that would probably not exist locally. Do we need some documentation around how to deploy it? Or maybe some best practices like creating an `/immutable` path that can be cached and images outside of that would not be cached or something? Maybe with examples for the built-in providers?

Author: Not sure I follow. What I meant is that if your `src` path starts with a `/` or with `http(s)://` it indicates that you're not importing some local image file on disk but instead want to reference something through the URL. What that means with regards to caching depends on where the image is stored. I wouldn't worry about that for now, we can document that later (since it's only a documentation thing, not an implementation issue).

Reviewer 2: using `/favicon.ico` resolves to `/static/favicon.ico` locally. so it could be confusing if some paths lead to local sources and others to a cdn url. Would an alias work? eg. $cdn where $cdn is a configured value from svelte/image/vite ?

Reviewer 3: problem solved

```suggestion
<Image src="https://example.com/path/to/your/image.jpg" width={1200} height={1800} alt="An alt text" />
```

```

> `width` and `height` should be the natural width/height of the referenced image. `alt` should describe the image. All are required. The `src` is transformed by calling `getURL` of the `default` provider provided in the `vite.config.js`.

Reviewer: ```suggestion
`width` and `height` should be the intrinsic width/height of the referenced image - i.e. the dimensions of the image before styling. `alt` should describe the image. All are required. The `src` is transformed by calling `getURL` of the `default` provider provided in the `vite.config.js`.
```

### Static build time optimization:

```svelte
<script>
	import Image from '@sveltejs/image';

>     import YourImage from '/path/to/your/image.jpg?generate';

Reviewer: I'd expect this to be provided by `defaultDirectives` as having to put it manually feels a bit clunky. Users should still be able to override it in the case. I'd also separate it into two pieces and use the standard directives. I'd make it `?w=100,200,400&as=img`. Then we can later have `?w=100,200,400&as=picture` (`as=img` will be built-in for the next major release). Doing it this way should also make it easy for users to override

Also, this line should be updated from spaces to tabs

</script>

<Image src={YourImage} alt="An alt text" />
```

> This optimizes the image at build time using `vite-imagetools`. `width` and `height` are optional as they can be infered from the source image.

Reviewer: ```suggestion
This optimizes the image at build time using `vite-imagetools`. `width` and `height` are optional as they can be inferred from the source image.
```

> ### Pros/Cons of the solutions

Reviewer: ```suggestion
### Choosing static vs dynamic
```

Using the static provider generates the images at build time, so that build time may take longer the more images you transform.

Using an image CDN provides more flexibility with regards to sizes and you can pass image sources not known at build time, but it comes with potentially a bit setup overhead (configuring the image CDN) and possibly usage cost.

You can mix and match both solutions in one project.

> ### `Image` component options

Reviewer: ```suggestion
### `Image` component props
```

There are a few things you can customize:

> - `priority`: give this to the most important/largest image on the page so it loads faster

Reviewer: clarify what to set it to
```suggestion
- `priority`: set this to `'high'` for the most important/largest image on the page so it loads faster
```

- `style`: to style the image
- `class`: to style the image. Be aware that you need to pass classes that are global (i.e. wrapped in `:global()` when coming from a `<style>` tag)
- `loading`: loading behavior of the image. Defaults to `lazy` which means it's loaded only when about to enter the viewport. Set to `eager` to load right away (the default when setting `priority`)
- `provider`: you can pass more than the `default` provider in `vite.config.js`. If you then want to use a different provider than the `default` one, pass it here
- `providerOptions`: provider-specific image CDN options. For example `quality` for Vercel

## Best practices

- Always provide a good `alt` text
- Your original images should have a good quality/resolution. Images are easier to size down than up
- Choose one image per page which is the most important/largest one and give it `priority` so it loads faster. This gives you better web vitals scores (largest contentful paint in particular)
- Give the image a container or a styling so that it is constrained and does not jump around. `width` and `height` help the browser reserving space while the image is still loading

> ## Roadmap

Reviewer: I didn't add the backticks around the code as it'd end the suggestion and I wasn't quite sure how to escape
```suggestion
## Automatic imports

You can use [svelte-preprocess-import-assets](https://github.com/bluwy/svelte-preprocess-import-assets) to automatically import your images:

import { vitePreprocess } from '@sveltejs/kit/vite';
import importAssets from 'svelte-preprocess-import-assets';

/** @type {import('@sveltejs/kit').Config} */
export default {
	preprocess: [
		importAssets({
			sources: (defaultSources) => {
				return [
					...defaultSources,
					{
						tag: 'Image',
						srcAttributes: ['src']
					}
				]
			}
		}),
		vitePreprocess()
	]
...

Then, rather than importing your image, you can simply write:
<Image src="./myimage.png" />
```

# `@sveltejs/image`

**WARNING**: This package is experimental. It uses pre-1.0 versioning and may introduce breaking changes with every minor version release.

This package aims to bring a plug and play image component to SvelteKit that is opinionated enough so you don't have to worry about the details, yet flexible enough for more advanced use cases or tweaks. It serves a smaller file format like `avif` or `webp` and uses the `srcset` and `sizes` attributes of the `img` tag to provide resized images suitable for various device sizes, which for example results in smaller images downloaded for mobile.

## Setup

Install:

```bash
npm install --save @sveltejs/image
```

Adjust `vite.config.js`:

```diff
+import { images } from '@sveltejs/image/vite';
import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vite';

export default defineConfig({
	plugins: [

> +		images({ providers: { default: '@sveltejs/image/providers/<choose one>' } }),

Reviewer: perhaps we could remove the `/providers/` portion as it's getting a bit long and things like `@sveltejs/image/vercel` or `@sveltejs/image/netlify` is pretty explanatory and won't conflict with anything else
```suggestion
+		images({ providers: { default: '@sveltejs/image/<choose one>' } }),
```

Reviewer 2: i like the explicit naming more. prevents conflicts too if future possible providers collided with a preexisting name.

How do you plan to have 3rd party providers added? PR here or is it possible to publish them separately eg `@cool/svelte-image-provider`

Author: Also vote for keeping the longer import. Any string that's a valid import is allowed there, you could even do a relative import (though we need to add support for that).

		sveltekit()
	]
});
```

> `<choose one>` refers to choosing one of the ready-to-use providers. We plan to add more providers over time. You can create your own by creating a JavaScript with a `export function getURL({ src, width, height }): string` function inside.

In case of Vercel, adjust `svelte.config.js`:

```diff
import adapter from '@sveltejs/adapter-vercel';

/** @type {import('@sveltejs/kit').Config} */
const config = {
	kit: {
-		adapter: adapter()

> +		adapter: adapter({ images: true })

Reviewer: is there a reason not to enable this by default? will anything bad happen if you do and don't have `@sveltejs/image` installed?

Author: Honestly I don't know, but it feels wrong to automatically enable it.

	}
};

export default config;
```

## Usage

### When using one of the providers, i.e. an image CDN:

```svelte
<script>
	import Image from '@sveltejs/image';
</script>

<Image src="/path/to/your/image.jpg" width={1200} height={1800} alt="An alt text" />
```

`width` and `height` should be the intrinsic width/height of the referenced image - i.e. the dimensions of the image before styling. `alt` should describe the image. All are required. The `src` is transformed by calling `getURL` of the `default` provider provided in the `vite.config.js`.

### Static build time optimization:

Install:

```bash
npm install --save-dev vite-imagetools
```

Use in your `.svelte` components:

```svelte
<script>

> 	import Image from '@sveltejs/image';

Reviewer: ```suggestion
	import { Image } from '@sveltejs/image';
```

</script>

<Image src="./path/to/your/image.jpg" alt="An alt text" />
```

This optimizes the image at build time using `vite-imagetools`. `width` and `height` are optional as they can be inferred from the source image.

> Note that the `src` must start with a `.`. `@sveltejs/image` includes a preprocessor that will convert it to the following:
>
> ```svelte
> <script>

> > 	import Image from '@sveltejs/image';

Reviewer: ```suggestion
> 	import { Image } from '@sveltejs/image';
```

> 	import YourImage from './path/to/your/image.jpg';
> </script>
>
> <Image src={YourImage} alt="An alt text" />
> ```

### Choosing static vs dynamic

> Using the static provider generates the images at build time, so that build time may take longer the more images you transform.

Reviewer: ```suggestion
Using the static provider generates the images at build time, so build time may take longer the more images you transform.
```

> Using an image CDN provides more flexibility with regards to sizes and you can pass image sources not known at build time, but it comes with potentially a bit setup overhead (configuring the image CDN) and possibly usage cost.

Reviewer: ```suggestion
Using an image CDN provides more flexibility with regards to sizes and you can pass image sources not known at build time, but it comes with potentially a bit of setup overhead (configuring the image CDN) and possibly usage cost.
```

- `provider`: you can pass more than the `default` provider in `vite.config.js`. If you then want to use a different provider than the `default` one, pass it here
- `providerOptions`: provider-specific image CDN options. For example `quality` for Vercel

## Best practices

- Always provide a good `alt` text
- Your original images should have a good quality/resolution. Images are easier to size down than up
- Choose one image per page which is the most important/largest one and give it `priority` so it loads faster. This gives you better web vitals scores (largest contentful paint in particular)
- Give the image a container or a styling so that it is constrained and does not jump around. `width` and `height` help the browser reserving space while the image is still loading

## Roadmap

This is an experimental MVP for getting initial feedback on the implementation/usability of an image component usable with SvelteKit (can also be used with Vite only). Once the API is stable, we'll want to create a more seamless integration with SvelteKit, i.e. less setup required.

## Acknowledgements

> We'd like to thank the authors of the Next/Nuxt/Astro/`unpic` image components for inspiring this work.

Reviewer: should we mention `svelte-preprocess-import-assets` and `vite-imagetools` or omit them since we're a bit involved with them?

Author: We definetly should also mention them!

# `@sveltejs/image`

**WARNING**: This package is experimental. It uses pre-1.0 versioning and may introduce breaking changes with every minor version release.

This package aims to bring a plug and play image component to SvelteKit that is opinionated enough so you don't have to worry about the details, yet flexible enough for more advanced use cases or tweaks. It serves a smaller file format like `avif` or `webp` and uses the `srcset` and `sizes` attributes of the `img` tag to provide resized images suitable for various device sizes, which for example results in smaller images downloaded for mobile.

## Setup

Install:

```bash
npm install --save @sveltejs/image
```

Adjust `vite.config.js`:

```diff
+import { images } from '@sveltejs/image/vite';
import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vite';

export default defineConfig({
	plugins: [

> +		images({ providers: { default: '@sveltejs/image/providers/<choose one>' } }),

Reviewer 3: this is wrong, it's missing the `runtime` portion:

```suggestion
+		images({
			runtime: {
				providers: {
					default: '@sveltejs/image/providers/<choose one>'
				}
			}
		}),
```

when you add it, the config starts to get _super_ unwieldy for the simple case. i do have to wonder if we really need multi-provider support, it's the edgiest of edge cases. If we _really, really_ needed it then maybe something like this would be neater?

```svelte
<script>
  const my_provider = {
    getSrc: ...
  };
</script>

<Image ... provider={my_provider} />
```

But to be clear I don't think that's something we need for an initial release. Let's just simplify the hell out of it

Author: How would the simple provider version look like then?

We also have to think about provider options which you might want to pass, like the default quality used for Vercel image optimization.

In case of Vercel, also adjust your `svelte.config.js`:

```diff
import adapter from '@sveltejs/adapter-vercel';

/** @type {import('@sveltejs/kit').Config} */
const config = {
	kit: {
-		adapter: adapter()
+		adapter: adapter({ images: true })
	}
};

export default config;
```

## Usage

### Dynamic runtime optimization using one of the providers, i.e. an image CDN:

```svelte
<script>
	import Image from '@sveltejs/image';
</script>

<Image src="/path/to/your/image.jpg" width={1200} height={1800} alt="An alt text" />
```

`width` and `height` should be the intrinsic width/height of the referenced image - i.e. the dimensions of the image before styling. `alt` should describe the image. All are required. The `src` is transformed by calling `getURL` of the `default` provider provided in the `vite.config.js`.

If you want to optimize an image from an external URL, add its domain to the `domains` config of the plugin.

### Static build time optimization:

> Static build time optimization uses `vite-imagetools`, which comes as an optional peer dependency, so you first need to install it:

Reviewer 3: I think we probably need some docs in this section about how installing `vite-imagetools` will change the behaviour of existing imports. In my app I had this...

```js
import logo from '$lib/icons/logo.avif';
```

```svelte
<img alt="logo" src={logo} />
```

...and it broke. (Ideally it wouldn't break in the first place, i.e. you'd have to explicitly opt in to `vite-imagetools` handling an import, but I'm not sure that's possible?)

Author: It would be possible if we expose our own imagetools directive, so people would do something like `import Foo from './my/image.jpg?as=svelteimage'` if they want to import it manually, and the preprocessor would add that automatically

```

You can also write it yourself like this, if you want to.

Note that the generated code is a `picture` tag (not an `img` tag as in the case of dynamic providers) wrapping one `source` tag per image type.

### Choosing static vs dynamic

Using the static provider generates the images at build time, so build time may take longer the more images you transform.

Using an image CDN provides more flexibility with regards to sizes and you can pass image sources not known at build time, but it comes with potentially a bit of setup overhead (configuring the image CDN) and possibly usage cost.

You can mix and match both solutions in one project.

### `Image` component props

There are a few things you can customize:

- `priority`: set this to `true` for the most important/largest image on the page so it loads faster
- `sizes`: If your image is less than full width on one or more screen sizes, add this info here. When using dynamic providers the widths can be adjusted accordingly to produce a more optimal `srcset`. See [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes) for more info on the attribute
- `style`: to style the image

> - `class`: to style the image. Be aware that you need to pass classes that are global (i.e. wrapped in `:global()` when coming from a `<style>` tag)

Author: Something like
```html
<img {...getImage({ src, width, height })} .. />
```
feels better to me at first. But it probably does not work with the static import preprocessor then..?

> - `class`: to style the image. Be aware that you need to pass classes that are global (i.e. wrapped in `:global()` when coming from a `<style>` tag)

Reviewer 3: (I don't think refactoring from one to the other is something we need to worry about — it's very unlikely that you'd need to do that, because they're really just different problems that happen to both involve an `<img>` as part of the solution)

> - `class`: to style the image. Be aware that you need to pass classes that are global (i.e. wrapped in `:global()` when coming from a `<style>` tag)

Reviewer: > I don't think refactoring from one to the other is something we need to worry about — it's very unlikely that you'd need to do that, because they're really just different problems that happen to both involve an <img> as part of the solution

I think a common case is that users will start with the build-time solution and then outgrow it and move to a CDN. The build-time solution is much easier to setup - it doesn't require setting up a CDN account, etc. But as sites grow I imagine many people will end up migrating to a CDN

> - `class`: to style the image. Be aware that you need to pass classes that are global (i.e. wrapped in `:global()` when coming from a `<style>` tag)

Reviewer 3: Build time to me means things like icons; run time means things like product shots, which you'd never have in your `src` folder in the first place (except maybe for some very early prototyping that you later replace with real data, which isn't the same as 'outgrowing')

> - `class`: to style the image. Be aware that you need to pass classes that are global (i.e. wrapped in `:global()` when coming from a `<style>` tag)

Reviewer 4: For a user perspective (indie SaaS & small startups), I'd prefer to stick with build-time generated images as long as practical speed wise--which could be a while for an indie SaaS with 3-4 marketing pages and low-volume blog. 

Being build-time static means 1.) fewer knobs to get off the ground, and 2.) better guarantees that a DDoS will cost nothing on hosts that provide zero-cost static file serving, which wouldn't be the case with a misconfigured cache, etc. 

A couple reasons to stay static as long as practical for some devs.
